README
NAME
    WWW::DreamHost::API - Perl interface to DreamHost Web Panel API

VERSION
    1.04

SYNOPSIS
     my $key = '6SHU5P2HLDAYECUM';
     my $api = new WWW::DreamHost::API($key);
     my $res = $api->command('api-list_accessible_cmds');
     print Data::Dumper->Dump([$res]);

DESCRIPTION
    WWW::DreamHost::API provides a simple Perl interface to DreamHost Web
    Panel API.

    Read more about API commands at
    <http://wiki.dreamhost.com/Application_programming_interface>

METHODS
  new ( $key )
    Creates WWW::DreamHost::API object

  uuid ( )
    Returns UUID. Probably actually unique (randomly based on key and run
    time). Don't worry, if it's not, because it's reinitializing in case of
    failure (though I cannot imagine how it can happen).

  reinit ( )
    If unique check fails, attempt to re-initialize. You don't have to call
    it yourself.

  command ( $cmd, [ $param => $value, ]+ )
    Execute a command.

    To get a list of availible commands, use something like that:

        my $res = $api->command('api-list_accessible_cmds');
        if ($res->{'success'}) {
            my @commands = @{ $res->{'data'} };
        }

    Returns a hash reference with (usually) 'result' and 'data' keys.
    'result' can be 'success' or 'error', and 'data' depends on command
    executed.

    See <http://wiki.dreamhost.com/Application_programming_interface> for
    more details.

CONFIGURATION AND ENVIRONMENT
    WWW::DreamHost::API is based on *libwww-perl* which uses *Crypt::SSLeay*
    or *IO::Socket::SSL*, with all consequences: HTTPS_proxy environment
    variable and so on. See *LWP* documentation unless you're already
    familiar with it.

DEPENDENCIES
    LWP::UserAgent Crypt::SSLeay Data::UUID LWP::protocol::https

    If JSON.pm is installed, JSON format is used when making requests to
    API; otherwise Data::Dumper format is used. Note that Data::Dumper
    format is "eval"-ed so (in theory) it can be used for security breach.

INCOMPATIBILITIES
    Not known.

BUGS AND LIMITATIONS
    Not known, but test suite MAY fail if DreamHost adds some other commands
    or change their behaviour. If you are using CPAN::Reporter in your
    service, I'll know about it. Consider installing it, as it really helps
    authors to know about possible bugs. See
    <http://wiki.cpantesters.org/wiki/QuickStart>.

AUTHOR
    Serguei Trouchelle stro@cpan.org

LICENSE AND COPYRIGHT
    This module is distributed under the same terms as Perl itself.

    Copyright (c) 2009-2012 Serguei Trouchelle



CHANGES
#   1.04    2012/03/24  If JSON is installed use it to prevent potential security breach (thanks to Darin McBride). Added "dynamic dns" script example.
#   1.03    2011/03/31  Added LWP::Protocol::https as prerequisite to resolve problems with libwww-perl v6
#   1.02    2009/11/22  Added "sanity check" test so at least one test is passed, regardless of presence of Test::More, Test::Pod::Coverage and Test::Pod -- to avoid UNKNOWN status
#   1.01    2009/10/18  Updated test prerequisites to Test::More 0.88 (thanks to Andreas Koenig)
#   1.00    2009/09/16  Initial revision  

eg/dyndns.pl
#!/usr/env perl

# $Id: dyndns.pl 38 2012-03-24 10:09:04Z stro $

use strict;
use warnings;
use Carp;
use English '-no_strict_refs';
                 
use Getopt::Long ();
use LWP::UserAgent;
use WWW::DreamHost::API;

use Data::Dumper;

my ($key, $domain);

my $opts = Getopt::Long::GetOptions(
    'key=s'    => \$key,
    'domain=s' => \$domain,
);


if (defined $key and defined $domain) {
    if (my $api = WWW::DreamHost::API->new($key)) {
        my $res = $api->command('api-list_accessible_cmds');
        die 'Cannot get a list of commands from API' unless $res->{'result'} eq 'success';

        my %commands = map { $_ => 1 } @{ $res->{'data'} };
#       die 'DNS commands looks disabled' unless $commands{'dns-add_record'} and $commands{'dns-remove_record'};

        # Check DNS address for old records

        $res = $api->command('dns-list_records');

        die 'Cannot get DNS records from API' unless $res->{'result'} eq 'success';

        foreach my $rec (@{ $res->{'data'} }) {
            if ($rec->{'record'} eq $domain) {
                next unless $rec->{'type'} eq 'A';

                die $domain, ' is not editable' unless $rec->{'editable'};

                # Remove old DNS address

                print 'Remove A record for ', $rec->{'value'}, "\n";

                $res = $api->command('dns-remove_record', 
                    'record' => $domain,
                    'type'   => 'A',
                    'value'  => $rec->{'value'},
                );

                die 'Cannot remove old A value' unless $res->{'result'} eq 'success';

                # last; if you're absolutely sure there's only one A record
            }
        }

        # Find what is our IP address

        my $ua = LWP::UserAgent->new;
        my $response = $ua->get('http://trouchelle.com/perl/ip.pl');

        die 'Cannot detect our IP' unless $response->is_success;

        my $addr = $response->decoded_content;

        # Create new A record

        $res = $api->command('dns-add_record', 
            'record' => $domain,
            'type'   => 'A',
            'value'  => $addr,
        );

        die 'Cannot create new A value' unless $res->{'result'} eq 'success';

        print 'New record for domain ', $domain, ' is ', $addr, "\n";

    } else {
        die 'Cannot create API object: ', $EVAL_ERROR;
    }
} else {
    print 'Usage: dyndns.pl key=6SHU5P2HLDAYECUM domain=dyn.trouchelle.com';
}



API.pm
package WWW::DreamHost::API;

# $Id: API.pm 38 2012-03-24 10:09:04Z stro $

use strict;
use warnings;
use Carp;

our $VERSION = '1.04';

use LWP::UserAgent;
use Data::UUID;

=head1 NAME

WWW::DreamHost::API - Perl interface to DreamHost Web Panel API

=head1 VERSION

1.04

=head1 SYNOPSIS

 my $key = '6SHU5P2HLDAYECUM';
 my $api = new WWW::DreamHost::API($key);
 my $res = $api->command('api-list_accessible_cmds');
 print Data::Dumper->Dump([$res]);

=head1 DESCRIPTION

WWW::DreamHost::API provides a simple Perl interface to DreamHost Web Panel API.

Read more about API commands at L<http://wiki.dreamhost.com/Application_programming_interface>

=head1 METHODS

=head2 new ( $key )

Creates WWW::DreamHost::API object

=cut

sub new {
    my $class = shift;
    my $key   = shift;

    croak 'You should provide a key in order to use WWW::DreamHost::API' unless defined $key;

    my $self = {
        '__key'  => $key,
        '__ua'   => LWP::UserAgent->new('agent' => 'WWW-DreamHost-API/' . $VERSION),
        '__ug'   => new Data::UUID,
        '__uniq' => int(rand(time())),

    };

    $self->{'__ua'}->env_proxy();

    # Check if JSON is available
    if (eval { require JSON; }) {
        $self->{'__format'} = 'json';
    } else {
        $self->{'__format'} = 'perl';
    }

    bless $self, $class;
    return $self;
}

=head2 uuid ( )

Returns UUID. Probably actually unique (randomly based on key and run time).
Don't worry, if it's not, because it's reinitializing in case of failure (though I cannot imagine how it can happen).

=cut

sub uuid {
    my $self = shift;
    my $uuid = $self->{'__ug'}->create_from_name('WWW-DreamHost-API-'. $self->{'__key'}, $self->{'__uniq'}++);
    return $self->{'__ug'}->to_string($uuid);
}

=head2 reinit ( )

If unique check fails, attempt to re-initialize. You don't have to call it yourself.

=cut

sub reinit {
    my $self = shift;
    $self->{'__uniq'} = int(rand(time()));
    return 1;
}

=head2 command ( $cmd, [ $param => $value, ]+ )

Execute a command. 

To get a list of availible commands, use something like that:

    my $res = $api->command('api-list_accessible_cmds');
    if ($res->{'success'}) {
        my @commands = @{ $res->{'data'} };
    }

Returns a hash reference with (usually) 'result' and 'data' keys. 'result' can be 'success' or 'error', and 'data' depends on command executed.

See L<http://wiki.dreamhost.com/Application_programming_interface> for more details.

=cut

sub command {
    my $self = shift;
    my $cmd = shift;
    my %extraparam = @_;
    delete $extraparam{$_} foreach (qw/ key cmd unique_id format /); # fool-proof

    while (1) {
        # Loop until UUID is unique. Though I'm VERY doubtful this can happen in real life.

        my $res = $self->{'__ua'}->post('https://api.dreamhost.com/', {
            'key'       => $self->{'__key'},
            'cmd'       => $cmd,
            'unique_id' => $self->uuid(),
            'format'    => $self->{'__format'},
            %extraparam,
        });

        if ($res->is_success()) {
            my $result;

            if ($self->{'__format'} eq 'json') {
                return unless $result = JSON::from_json($res->content());
            } else {
                return unless eval $res->content();
            }

            return unless ref($result) eq 'HASH';

            if ($result->{'result'} eq 'error' and $result->{'data'} eq 'unique_id_already_used') {
                $self->reinit();    # Reinitialize random seed
                redo;               # Send another request
            }

            return $result;
        } else {
            eval { die $res->status_line(); };
            return;
        }
    }

    return; # for Perl::Critic
}

=head1 CONFIGURATION AND ENVIRONMENT

WWW::DreamHost::API is based on I<libwww-perl> which uses I<Crypt::SSLeay>
or I<IO::Socket::SSL>, with all consequences: HTTPS_proxy environment
variable and so on. See I<LWP> documentation unless you're already familiar
with it.

=head1 DEPENDENCIES

LWP::UserAgent
Crypt::SSLeay
Data::UUID
LWP::protocol::https

If JSON.pm is installed, JSON format is used when making requests to API;
otherwise Data::Dumper format is used. Note that Data::Dumper format is
"eval"-ed so (in theory) it can be used for security breach. 

=head1 INCOMPATIBILITIES

Not known.

=head1 BUGS AND LIMITATIONS

Not known, but test suite MAY fail if DreamHost adds some other commands or
change their behaviour.
If you are using L<CPAN::Reporter> in your service, I'll know about it.
Consider installing it, as it really helps authors to know about possible bugs.
See L<http://wiki.cpantesters.org/wiki/QuickStart>.

=head1 AUTHOR

Serguei Trouchelle L<stro@cpan.org>

=head1 LICENSE AND COPYRIGHT

This module is distributed under the same terms as Perl itself.

Copyright (c) 2009-2012 Serguei Trouchelle

=cut

1;